Feature/full function

asv-hungvt opened the pull request:
add function edit and remove when user logins

asv-hungvt committed:
add funcion edit and delete comments with different users

asv-hungvt committed:
fix edit form to show content the form

vn-trinhtk commented on djangoappengine/guestbook/forms.py:
```diff
@@ -1,4 +1,36 @@
+    greeting_author = forms.CharField(label="Author", required=False, widget=forms.TextInput(attrs={'readonly': 'readonly'}))
+    greeting_content = forms.CharField(label="", required=True, max_length=10, widget=forms.Textarea)
+
+    def update_greeting(self):
```
thao tác update, delete, dời vào model.

asv-hungvt replied:
cái này không dời được vì hàm này là hàm của form, sẽ được gọi bởi tham số form trong hàm form_valid 
phần này chỉ truyền tham số, việc thao tác dữ liệu vẫn nằm ở model

vn-trinhtk commented on djangoappengine/guestbook/models.py:
```diff
@@ -68,4 +68,21 @@ def put_greeting(self, content, author, user, title):
+
+
+    def update_greeting_by_id(self, author, user, is_superuser, id, greeting_content):
+        obj = ndb.Key('Guestbook', 'default_guestbook', Greeting, int(id)).get()
```
không hardcode `default_guestbook`, dùng attr name

vn-trinhtk commented on djangoappengine/guestbook/models.py:
```diff
@@ -68,4 +68,21 @@ def put_greeting(self, content, author, user, title):
+    def update_greeting_by_id(self, author, user, is_superuser, id, greeting_content):
+        obj = ndb.Key('Guestbook', 'default_guestbook', Greeting, int(id)).get()
+        obj.content = greeting_content
+        obj.put()
```
lỡ `put` failed cũng xoá memcache ?

vn-trinhtk commented on djangoappengine/guestbook/templates/guestbook/main_page.html:
```diff
@@ -10,6 +10,28 @@
+                <input type="hidden" name="greeting_author" value="{{ greeting.author }}" />
+                <input type="hidden" name="greeting_content" value="" />
+                <input type="hidden" name="guestbook_name" value="{{ guestbook_name }}" />
+                <div><input type="submit" value="Edit"></div>
```
không cần submit như vậy: 

```
<a href="action/?guestbook_name={{ guestbook_name }}&id={{ greeting.key.id }}...">Action</a>
```

vn-trinhtk commented on djangoappengine/guestbook/views.py:
```diff
@@ -62,11 +62,59 @@ def form_invalid(self, form):
+        author = None
+        if users.get_current_user():
+            author = users.get_current_user()
+        form.delete_message_form()
```
`guestbook` delete greeting, ko phải form làm thao tác này

asv-hungvt replied:
phần này giống như phần comment này:
https://github.com/asv-rak/training-python-hungvt/pull/6#discussion_r74029206

vn-trinhtk commented on djangoappengine/guestbook/views.py:
```diff
@@ -62,11 +62,59 @@ def form_invalid(self, form):
+        return self.render_to_response(self.get_context_data(form=my_form))
+
+
+class DeleteFormView(FormView):
```
Delete chỉ có thao tác xoá, có cần phải dùng FormView ko ?

asv-hungvt replied:
Do lúc delete cần id truyền vào và hiện tại anh thấy việc dùng form sẽ giúp mình lấy đc id nên anh làm theo cách này, 
Mình có cách khác để làm hả Trình?

asv-hungvt committed:
- change post method to get method in loading edit form
- remove hardcoding guestbook name

asv-hungvt committed:
remove hardcoding

asv-hungvt committed:
update view without using form m

asv-hungvt committed:
remove unnecessary code in form

asv-hungvt committed:
- change django version to make it adapt with GAE
- add api for getting mesage detail
- add conventional model functions

asv-hungvt committed:
- add function get all items from a guestbook

asv-hungvt committed:
- da them chuc nang post mot message moi
- hien thoi anh dang disable csrf khi goi method post, anh se tim hieu ve viec add them csrf khi gui request post vao t2

asv-hungvt committed:
add put and delete method API

asv-hungvt committed:
fix return code and return result from model

asv-hungvt committed:
add csrf token when submitting a f

asv-hungvt committed:
fix comments and backup before chaging m

asv-hungvt committed:
finish updating c

asv-hungvt committed:
add comments

asv-hungvt committed:
removed unused URL

asv-hungvt committed:
remove print method

asv-hungvt committed:
add cursor, investigating why cursor couldn't fetch next

asv-hungvt committed:
add pagination in get method

asv-hungvt committed:
remove unsued code

asv-hungvt committed:
delete unused code

asv-hungvt committed:
add exception message

asv-hungvt committed:
remove unused var

Commit:
Merge pull request #7 from asv-rak/feature/api

Feature/api
